sql/2024/privacy/ara-trigger-registrations-for-different-destinations-by-destinations.sql

```diff
@@ -0,0 +1,78 @@
+-- ara-trigger-registrations-for-different-destinations-by-destinations.sql
+-- Analysis of Attribution Reporting API (ARA) Triggers registered for different destinations by destination domains (i.e., advertiser domains):
+-- 1. No. of third-parties that register trigger for the given destination
+-- 2. Min. epsilon -- MIN(CASE WHEN epsilon IS NOT NULL THEN epsilon END) AS min_epsilon
+-- 3. Avg. epsilon -- AVG(CASE WHEN epsilon IS NOT NULL THEN epsilon END) AS avg_epsilon
+-- 4. Max. epsilon -- MAX(CASE WHEN epsilon IS NOT NULL THEN epsilon END) AS max_epsilon
+-- [Higher the epsilon, the more the privacy protection] [Epsilon is always undefined, so last 3 columns are removed for this year atleast]
+-- Output comprises 9.5K rows and 1 column.
+
+-- Extracting third-parties observed using ARA API on a publisher
+CREATE TEMP FUNCTION jsonObjectKeys(input STRING)
+RETURNS ARRAY<STRING>
+LANGUAGE js AS """
+  if (!input) {
+    return [];
+  }
+  return Object.keys(JSON.parse(input));
+""";
+
+-- Extracting ARA API source registration details being passed by a given third-party (passed as "key")
+CREATE TEMP FUNCTION jsonObjectValues(input STRING, key STRING)
+RETURNS ARRAY<STRING>
+LANGUAGE js AS """
+  if (!input) {
+    return [];
+  }
+  const jsonObject = JSON.parse(input);
+  const values = jsonObject[key] || [];
+  const result = [];
+
+  values.forEach(value => {
+    if (value.toLowerCase().startsWith('attribution-reporting-register-source|')) {
+      const parts = value.replace('attribution-reporting-register-source|', '').split('|');
+      parts.forEach(part => {
+        if (part.startsWith('destination=')) {
+          const destinations = part.replace('destination=', '').split(',');
+          destinations.forEach(destination => {
+            result.push('destination=' + destination.trim());
+          });
+        } else {
+          result.push(part.trim());
+        }
+      });
+    }
+  });
+
+  return result;
+""";
+
+WITH ara_features AS (
+  SELECT
+    NET.REG_DOMAIN(page) AS publisher,
+    third_party_domain,
+    CASE
+      WHEN ara LIKE 'destination=%' THEN NET.REG_DOMAIN(REPLACE(ara, 'destination=', ''))
+      ELSE NULL
+    END AS destination,
+    CASE
+      WHEN ara LIKE 'epsilon=%' THEN SAFE_CAST(REPLACE(ara, 'epsilon=', '') AS FLOAT64)
+      ELSE NULL
+    END AS epsilon
+  FROM `httparchive.all.pages`,
+    UNNEST(jsonObjectKeys(JSON_QUERY(custom_metrics, '$.privacy-sandbox.privacySandBoxAPIUsage'))) AS third_party_domain,
+    UNNEST(jsonObjectValues(JSON_QUERY(custom_metrics, '$.privacy-sandbox.privacySandBoxAPIUsage'), third_party_domain)) AS ara
+  WHERE
+    date = '2024-06-01' AND
+    client = 'desktop' AND
+    is_root_page = TRUE AND
+    rank <= 1000000
```

Comment:
@Yash-Vekaria can we remove this filter for production queries, or what is your idea?

Reply:
The idea was just to analyse top 1M websites (also because full data might be too much). But, I'm fine removing it. 

Also, all my SQLs are written to produce the full output because I'm not sure what specific metric we might want to be included in the chapter to be discussed. But, I don't know how future stages will be performed. As in during chapter writing, will authors tell us what they want to include, and we should compute that and provide them based on the output generated from running these queries (however, queries are not meant to write generated output to some table that can be later re-used to compute analysis for different things).

Reply:
See [the analyst's workflow](https://github.com/HTTPArchive/almanac.httparchive.org/wiki/Analysts'-Guide#analysis-workflow).
We need to provide the charts with the insights you want to showcase for the chapter in [2024 Chapter Sheets](https://docs.google.com/spreadsheets/d/18r8cT6x9lPdM-rXvXjsqx84W7ZDdTDYGD59xr0UGOwg/edit).
The authors will be able to insert the charts into the chapter and write about the insights.

So write your queries with the particular charts in mind.

Example from 2022: [query](https://github.com/HTTPArchive/almanac.httparchive.org/blob/main/sql/2022/privacy/number_of_websites_with_privacy_service.sql) & [sheet with a chart](https://docs.google.com/spreadsheets/d/1iJqj3g0VEjpmjzvtX6VLeRehE7LDQGcw6lOadxGxkjk/edit?gid=988200988#gid=988200988).

Reply:
Ok I will revise my queries (probably it will increase the number of sql files). Once done, will push them. Also, remove the rank filters from all the queries. 

Can you grant edit access to the metrics/resulta sheet?